File: src/main/java/com/example/the_machine/domain/EventLogEntity.java
package com.example.the_machine.domain;

import com.fasterxml.jackson.databind.JsonNode;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.Instant;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

@Entity
@Table(name = "event_log")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class EventLogEntity {

  @Id
  private UUID id;

  @Column(nullable = false)
  private UUID threadId;

  private UUID runId;

  private UUID messageId;

  @Enumerated(EnumType.STRING)
  @Column(nullable = false)
  private EventType eventType;

  @Column(nullable = false, columnDefinition = "jsonb")
  @JdbcTypeCode(SqlTypes.JSON)
  private JsonNode payload;

  @Column(nullable = false)
  private Instant occurredAt;

  private String dedupeKey;

}

File: src/main/kotlin/com/example/the_machine/domain/EventLogEntity.kt
package com.example.the_machine.domain

import com.example.the_machine.common.JsonElementConverter
import jakarta.persistence.Column
import jakarta.persistence.Convert
import jakarta.persistence.Entity
import jakarta.persistence.Table
import kotlinx.serialization.json.JsonElement
import org.hibernate.annotations.JdbcTypeCode
import org.hibernate.type.SqlTypes
import java.time.Instant
import java.util.*

@Entity
@Table(name = "event_log")
data class EventLogEntity(

  @Column(nullable = false)
  val threadId: UUID,

  val runId: UUID?,

  val messageId: UUID?,

  @Column(nullable = false)
  val eventType: EventType,

  @Column(nullable = false, columnDefinition = "jsonb")
  @JdbcTypeCode(SqlTypes.JSON)
  @Convert(converter = JsonElementConverter::class)
  val payload: JsonElement,

  @Column(nullable = false)
  val occurredAt: Instant,

  val dedupeKey: String?
) : BaseEntity()